src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/utils/SubQuery_Analyzer.java

```diff
@@ -0,0 +1,72 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils;
+
+import org.apache.jena.graph.Node;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpMultiwayUnion;
+import se.liu.ida.hefquin.engine.queryplan.utils.LogicalOpUtils;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpRequest;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpUnion;
+
+import java.util.*;
+
+public class SubQuery_Analyzer {
+    private List<Node> subs, preds, objs;
+
+    public SubQuery_Analyzer( final LogicalPlan lop ) {
+        subs = new ArrayList<>();
+        preds = new ArrayList<>();
+        objs = new ArrayList<>();
+
+        analyze(lop);
```

Comment:
I wouldn't hide the functionality to populate the three lists inside that function because that makes the code less easy to understand. Instead, use the function only to extract/obtain the set of triple patterns and, then, populate the lists directly here in the constructor.